services/libs/aster-std/src/syscall/set_get_priority.rs

```diff
@@ -0,0 +1,124 @@
+    log_syscall_entry!(SYS_GET_PRIORITY);
+    let prio_target = PriorityTarget::new(which, who)?;
+    debug!("get_priority prio_target: {:?}", prio_target);
+
+    let processes = get_processes(prio_target)?;
+    let highest_prio = {
+        let mut nice = Nice::lowest_priority();
+        for process in processes.iter() {
+            let proc_nice = process.nice().lock();
+            // Returns the highest priority enjoyed by the processes
+            if *proc_nice < nice {
+                nice = *proc_nice;
+            }
+        }
+
+        // The system call returns nice values translated to the range 40 to 1,
+        // since a negative return value would be interpreted as an error.
+        RlimitNice::from(nice)
+    };
+
+    Ok(SyscallReturn::Return(highest_prio.to_raw() as _))
+}
+
+fn get_processes(prio_target: PriorityTarget) -> Result<Vec<Arc<Process>>> {
+    Ok(match prio_target {
+        PriorityTarget::Process(pid) => {
+            let process = process_table::get_process(&pid).ok_or(Error::new(Errno::ESRCH))?;
+            vec![process]
+        }
+        PriorityTarget::ProcessGroup(pgid) => {
+            let process_group =
+                process_table::get_process_group(&pgid).ok_or(Error::new(Errno::ESRCH))?;
+            let processes = process_group.get_all_processes();
+            if processes.is_empty() {
+                return_errno!(Errno::ESRCH);
+            }
+            processes
+        }
+        PriorityTarget::User(uid) => {
+            let processes = process_table::get_user_processes(&uid);
+            if processes.is_empty() {
+                return_errno!(Errno::ESRCH);
+            }
+            processes
+        }
+    })
+}
+
+#[derive(Debug)]
+enum PriorityTarget {
+    Process(Pid),
+    ProcessGroup(Pgid),
+    User(Uid),
+}
+
+impl PriorityTarget {
+    fn new(which: i32, who: u32) -> Result<Self> {
+        let which = c_prio_which::try_from(which)
+            .map_err(|_| Error::with_message(Errno::EINVAL, "invalid which value"))?;
+        Ok(match which {
+            c_prio_which::PRIO_PROCESS => {
+                let pid = if who == 0 {
+                    current!().pid()
+                } else {
+                    who as Pid
+                };
+                Self::Process(pid)
+            }
+            c_prio_which::PRIO_PGRP => {
+                let pgid = if who == 0 {
+                    current!().pgid()
+                } else {
+                    who as Pgid
+                };
+                Self::ProcessGroup(pgid)
+            }
+            c_prio_which::PRIO_USER => {
+                let uid = if who == 0 {
+                    credentials().ruid()
+                } else {
+                    Uid::new(who)
+                };
+                Self::User(uid)
+            }
+        })
+    }
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Debug, TryFromInt)]
+#[repr(i32)]
+enum c_prio_which {
```

Comment:
Sometimes we choose lower-case names to be compatible with the C definition. But there is not a C type called `c_prio_which`. So in this specific case, using the lower-case name is not wise. I suggest calling it `Which`. It is ok to give a short name as it is an internal type only used in this module. So the short name won't cause any ambiguity.

Reply:
OK